xml/value.go

```diff
@@ -0,0 +1,339 @@
+package xml
+
+import (
+	"encoding/base64"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+	"strings"
+)
+
+// Value represents an XML Value type
+// XML Value types: Object, Array, Map, String, Number, Boolean, and Null.
+type Value struct {
+	w       writer
+	scratch *[]byte
+
+	startElement *StartElement
+	endElement   *EndElement
+}
+
+// newValue returns a new Value encoder
+func newValue(w writer, scratch *[]byte, startElement *StartElement, endElement *EndElement) Value {
+	return Value{
+		w:            w,
+		scratch:      scratch,
+		startElement: startElement,
+		endElement:   endElement,
+	}
+}
+
+// writeStartElement takes in a start element and writes it.
+// It handles namespace, attributes in start element.
+func writeStartElement(w writer, el *StartElement) error {
+	if el == nil {
+		return nil
+	}
+
+	w.WriteRune(leftAngleBracket)
+
+	if len(el.Name.Space) != 0 {
+		w.WriteString(el.Name.Space)
+		w.WriteRune(colon)
+	}
+	w.WriteString(el.Name.Local)
+
+	for _, attr := range el.Attr {
+		// if attribute name len is zero or name.space is `xmlns`, it is a namespace attribute
+		if len(attr.Name.Local) == 0 || strings.EqualFold(attr.Name.Space, "xmlns") {
+			w.WriteRune(' ')
+			buildNamespace(w, &attr)
```

Comment:
What's the reasoning for having this distinction of seperating out attributes with namespace versus no namespace?

Reply:
Looking a bit further I think we might be overloading this a bit, from how I understand this, the case you are trying to handle here is where we want to just write out xmlns="uriFromSmithy". The specification seems to indicate that this declaration is just a attribute name not a namespace prefix without an attribute name. https://www.w3.org/TR/REC-xml-names/#NT-DefaultAttName To be it would be simpler to just combine buildNamespace and buildAttribute into a single function as they are essentially the same thing. Smithy codegen should pass the URI namespace as a local name xmlns, and not a namespace without a local name.